infra/Pos/Slotting/MemState/Class.hs

```diff
@@ -8,24 +8,52 @@ import           Control.Monad.Trans (MonadTrans)
 import           Universum
 
 import           Pos.Core.Types      (EpochIndex, Timestamp)
-import           Pos.Slotting.Types  (SlottingData)
+import           Pos.Slotting.Types  (EpochSlottingData)
+
 
 -- | 'MonadSlotsData' provides access to data necessary for slotting to work.
+-- Both _current_ and _next_ epoch @SlottingData@ need to be present since it's impossible
+-- to create @SlottingData@ without them.
 class Monad m => MonadSlotsData m where
 
-    getSystemStart :: m Timestamp
+    getSystemStartM :: m Timestamp
+
+    getAllEpochIndicesM :: m [EpochIndex]
+
+    getCurrentEpochIndexM :: m EpochIndex
+
+    getNextEpochIndexM :: m EpochIndex
 
-    getSlottingData :: m SlottingData
+    getCurrentEpochSlottingDataM :: m EpochSlottingData
 
-    waitPenultEpochEquals :: EpochIndex -> m ()
+    getNextEpochSlottingDataM :: m EpochSlottingData
 
-    putSlottingData :: SlottingData -> m ()
+    getEpochSlottingDataM :: EpochIndex -> m (Maybe EpochSlottingData)
+
+    putEpochSlottingDataM :: EpochIndex -> EpochSlottingData -> m ()
+
+    waitCurrentEpochEqualsM :: EpochIndex -> m ()
```

Comment:
1. https://github.com/input-output-hk/cardano-sl/pull/1217#issuecomment-318352696 under 3.
2. I agree I can add some comments but I wouldn't say I figured out how they work since there is still a lot of things I'm missing.
3. Reading your comments, you seem to be enjoying very dense code. Which is an aesthetic preference you have every right to. I found it **more readable**, and you can look at the answer here - https://softwareengineering.stackexchange.com/questions/17305/how-do-you-use-blank-lines-in-your-code
Of course, I can return it if you insist, but you might be chasing LOC number more than readability?